### PR TITLE
List a tiny JS world by Lu-sure

### PR DESCRIPTION
**Pull request by Lu-sure**
### A tiny JS world

[Demo](https://lu-sure.github.io/A_tiny_JS_world/) | [Code base](https://github.com/Lu-sure/Lu-sure.github.io/tree/main/A_tiny_JS_world)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/Lu-sure/index.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+   Code repository: https://github.com/Lu-sure/Lu-sure.github.io/tree/main/A_tiny_JS_world
+   Web app: https://lu-sure.github.io/A_tiny_JS_world/
+   */
+
+// ======== OBJECTS DEFINITIONS ========
+const dog = {
+   species: 'dog',
+   name: 'Easy',
+   gender: 'female',
+   age: '2',
```
Comment (OleksiyRudenko) [inline submissions/Lu-sure/index.js:13]: Why age is a string?

File: submissions/Lu-sure/index.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+   legs: 4,
+   hands: 0,
+   saying: {text: 'woof-woof!'},
```
Comment (OleksiyRudenko) [inline submissions/Lu-sure/index.js:16]: Why do we need an extra nested object here instead of just plain string?

File: submissions/Lu-sure/index.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+const inhabitants = [dog, cat, man, woman, catWoman];
+
+const getPrintText = (obj) => Object.values(obj)
+  .reduce((acc, value) => {
+      const perfix = acc !== '' ? '; ' : '';
+      let text = value;
+      if (value != null && !Array.isArray(value) && typeof value === 'object') text = getPrintText(value);
+      return `${acc}${perfix}${text}`;
+   }, '');
```
Comment (OleksiyRudenko) [inline submissions/Lu-sure/index.js:72]: `getPrintText` looks a bit overcomplicated. Seems like there is some room for optimization by means of `Array.map` and `Array.join`. Try making code in this function cleaner.

File: submissions/Lu-sure/index.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+   friendly: 'Roxy, Kolia'
+ };
+
+const inhabitants = [dog, cat, man, woman, catWoman];
+
+const getPrintText = (obj) => Object.values(obj)
+  .reduce((acc, value) => {
+      const perfix = acc !== '' ? '; ' : '';
+      let text = value;
+      if (value != null && !Array.isArray(value) && typeof value === 'object') text = getPrintText(value);
```
Comment (OleksiyRudenko) [inline submissions/Lu-sure/index.js:70]: Why do we need to test `value` against `null` and `Array` type?

File: submissions/Lu-sure/index.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      return `${acc}${perfix}${text}`;
+   }, '');
+
+// ======== OUTPUT ========
+for (let i=0; i<inhabitants.length; i++) {
+   print(getPrintText(inhabitants[i]));
+};
```
Comment (OleksiyRudenko) [inline submissions/Lu-sure/index.js:77]: Please employ `Array.forEach` here

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@Lu-sure nicely done.
There is a couple of questions below. Please answer, i am curious about motivation behind the implementation decisions taken.
Please also fix as per comments.
Feel free asking questions.

**Comment by OleksiyRudenko** [comment]
Please fix your submissions file structure.
What we've got here is:
image

While we expect that each task also has its own individual folder. Otherwise you end up having all submissions mixed.
image

File: submissions/Lu-sure/index.js (hunk 1 above)
Comment (Lu-sure) [reply submissions/Lu-sure/index.js:13]: no reason, you're right))

File: submissions/Lu-sure/index.js (hunk 5 above)
Comment (Lu-sure) [reply submissions/Lu-sure/index.js]: Done! Thanks))

File: submissions/Lu-sure/index.js (hunk 3 above)
Comment (Lu-sure) [reply submissions/Lu-sure/index.js]: Totally agree, done.

File: submissions/Lu-sure/index.js (hunk 2 above)
Comment (Lu-sure) [reply submissions/Lu-sure/index.js]: Obviously, something gone wrong during writing this . And the article about copying objects confused me. Fixed that already)

File: submissions/Lu-sure/tiny-JS-world/index.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  };
+ 
+  const man = {
+    species: 'human',
+    name: 'Kolia',
+    gender: 'male',
+    age: 26,
+    legs: 2,
+    hands: 2,
+    saying: 'Let`s work all day long!',
```
Comment (OleksiyRudenko) [inline submissions/Lu-sure/tiny-JS-world/index.js:37]: Check [special characters escaping](https://www.w3schools.com/js/js_strings.asp) to avoid situations where you use a **similar** character (`` ` ``) in place of the proper one (`'`).
**You do not need to fix this** here, but make sure that you will address similar situations properly in your future submission.

File: submissions/Lu-sure/tiny-JS-world/index.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+ 
+  const catWoman = {
+    species: 'cat+human',
+    name: 'Lila',
+    gender: 'female',
+    age: 21,
+    legs: 2,
+    hands: 2,
+    saying: cat.saying,
+    friendly: 'Roxy, Kolia'
```
Comment (OleksiyRudenko) [inline submissions/Lu-sure/tiny-JS-world/index.js:60]: A bit too straightforward approch. List of friends should be a list.
**You do not need to fix this** here, but make sure that you will address similar situations properly in your future submission.

File: submissions/Lu-sure/tiny-JS-world/index.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+  };
+ 
+ const inhabitants = [dog, cat, man, woman, catWoman];
+ 
+ const getPrintText = (obj) => Object.values(obj).join('; ');
```
Comment (OleksiyRudenko) [inline submissions/Lu-sure/tiny-JS-world/index.js:65]: This fragment relies on that `Object.values` returns object properties' values in a particular order. This is not guaranteed by design. Besides, explicit list of properties to print is better on a scale.
**You do not need to fix this** here, but make sure that you will address similar situations properly in your future submission.
Hint `['propA', 'propB'].map(propName => obj[propName])`

File: submissions/Lu-sure/tiny-JS-world/index.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+ 
+ inhabitants.forEach(creature => print(getPrintText(creature)));
+
```
Comment (OleksiyRudenko) [inline submissions/Lu-sure/tiny-JS-world/index.js:68]: This symbol
image
normally means there is no empty line at the end of file.

Here we can see line 68 is empty yet github warns something is wrong with the end of file.
Possible reasons:
(1) line 68 contains some whitespaces and technically this is not an empty line
(2) newline chatacter does not contain LF (Unix style newline, and a part of Windows newline style CRLF), which may happen when the platform is Mac (CR newline character) and/or git newline setting is not set properly

You may want to check your environment based on [this](https://gist.github.com/OleksiyRudenko/d51388345ea55767b7672307fe35adf3) and set proper [newlines handling by git](https://docs.github.com/en/free-pro-team@latest/github/using-git/configuring-git-to-handle-line-endings)

**You do not need to fix this** here, but make sure to tune your IDE and git properly to avoid the issue in your future submissions.

**Review by OleksiyRudenko — APPROVED** [review]
@Lu-sure well done! Approved.
Please check the comments below for the sake of your future submissions code and developer's environment appropriateness.
Fell fre asking questions if any.

File: submissions/Lu-sure/tiny-JS-world/index.js (hunk 9 above)
Comment (Lu-sure) [reply submissions/Lu-sure/tiny-JS-world/index.js:68]: I found out the actual reason is (1), the line contains whitespase. Thank u))

File: submissions/Lu-sure/tiny-JS-world/index.js (hunk 6 above)
Comment (Lu-sure) [reply submissions/Lu-sure/tiny-JS-world/index.js:37]: Yes, you are right. Was unobservant..

File: submissions/Lu-sure/tiny-JS-world/index.js (hunk 7 above)
Comment (Lu-sure) [reply submissions/Lu-sure/tiny-JS-world/index.js:60]: Do you mean array of friends?

Comment (OleksiyRudenko) [reply submissions/Lu-sure/tiny-JS-world/index.js:60]: Yes

File: submissions/Lu-sure/tiny-JS-world/index.js (hunk 9 above)
Comment (OleksiyRudenko) [reply submissions/Lu-sure/tiny-JS-world/index.js:68]: This is also auto-fixable with proper IDE settings. Make IDE do dirty job for you.

File: submissions/Lu-sure/tiny-JS-world/index.js (hunk 8 above)
Comment (Lu-sure) [reply submissions/Lu-sure/tiny-JS-world/index.js:65]: I finally figured it out)) Thanks!

File: submissions/Lu-sure/tiny-JS-world/index.js (hunk 9 above)
Comment (Lu-sure) [reply submissions/Lu-sure/tiny-JS-world/index.js:68]: Already made! Good idea)
